src/D3MHub.sol

```diff
@@ -219,170 +219,123 @@ contract D3MHub {
-        // This module will have an unintended behaviour if rate is changed to some other value.
-
-        EndLike _end;
-        uint256 daiDebt;
-        if (mode == Mode.NORMAL) {
-            // Normal mode or module just caged (no culled)
-            // debt is obtained from CDP art
-            (, daiDebt) = vat.urns(ilk, address(_pool));
-        } else if (mode == Mode.D3M_CULLED) {
-            // Module shutdown and culled
-            // debt is obtained from free collateral owned by the pool contract
-            // We rebalance the CDP after grabbing in `cull` so the gems represents
-            // the debt at time of cull
-            daiDebt = vat.gem(ilk, address(_pool));
-        } else if (mode == Mode.MCD_CAGED) {
-            // MCD caged
-            // debt is obtained from free collateral owned by the End module
-            _end = end;
-            _end.skim(ilk, address(_pool));
-            daiDebt = vat.gem(ilk, address(_end));
-        } else revert("D3MHub/unknown-mode");
-
-        uint256 availableAssets = _pool.maxWithdraw();
-
-        // Unwind amount is limited by how much:
-        // - max reduction desired
-        // - assets available
-        // - dai debt tracked in vat (CDP or free)
-        uint256 amount = _min(
-                            _min(
-                                supplyReduction,
-                                availableAssets
-                            ),
-                            daiDebt
-                        );
-        require(amount <= MAXINT256, "D3MHub/overflow");
-
-        // Determine the amount of fees to bring back
-        uint256 fees = 0;
-        if (assetBalance > daiDebt) {
+    function _exec(bytes32 ilk, ID3MPool _pool, uint256 Art, uint256 lineWad) internal {
+        require(lineWad <= SAFEMAX, "D3MHub/lineWad-above-max-safe");
+        (uint256 ink, uint256 art) = vat.urns(ilk, address(_pool));
+        require(ink <= SAFEMAX, "D3MHub/ink-above-max-safe");
+        require(ink >= art, "D3MHub/ink-not-greater-equal-art");
+        require(art == Art, "D3MHub/more-than-one-urn");
+        uint256 currentAssets = _pool.assetBalance(); // Should return DAI owned by D3MPool
+        uint256 maxWithdraw = _min(_pool.maxWithdraw(), SAFEMAX);
+
+        // Determine if fees were generated and try to account them (or the most that it is possible)
+        if (currentAssets > ink) {
+            uint256 fixInk = _min(
+                _min(
+                    currentAssets - ink, // fees generated
+                    ink < lineWad // if previously CDP was under debt ceiling
+                        ? (lineWad - ink) + maxWithdraw // up to gap to reach debt ceiling + maxWithdraw
+                        : maxWithdraw // up to maxWithdraw
+                ),
+                SAFEMAX + art - ink //  ensures that fixArt * RAY (rate) will be <= MAXINT256 (in vat.grab)
```

Comment:
Not sure I understand the comment.
Isn't `fixArt * RAY <= MAXINT256` `because `fixArt = ink - art` and `ink < SAFEMAX`?
Maybe the comment should be about why the min calculation includes `SAFEMAX + art - ink`.

Reply:
So we need to make sure `fixArt * RAY` is less than `MAXINT256` as otherwise in `vat.grab`, this line `int dtab = mul(ilk.rate, dart);` will fail.
It's true that `ink` is `<= SAFEMAX`, however `ink` can previously be increased by `currentAssets`. So the whole idea of including this part in the `min` is to make sure that the end resulted `fixArt`, which will be ` >= fixInk` doesn't overflow.
Not sure how to make the comment clearer though.